km/km_management.c

```diff
@@ -184,3 +189,149 @@ void km_mgt_init(char* path)
+      return -1;
+   }
+
+   cb_sockaddr = (struct sockaddr*)malloc(info->ai_addrlen);
+   memcpy(cb_sockaddr, info->ai_addr, info->ai_addrlen);
+   cb_sockaddr_len = info->ai_addrlen;
+
+   freeaddrinfo(info);
+   return 0;
+}
+
+static int km_build_unixaddr(char* path)
+{
+   struct sockaddr_un* addr = (struct sockaddr_un*)malloc(sizeof(struct sockaddr_un));
+   if (strlen(path) >= sizeof(addr->sun_path)) {
+      free(addr);
+      cb_sockaddr = NULL;
+      return -1;
+   }
+   strncpy(addr->sun_path, path, sizeof(addr->sun_path));
+   addr->sun_family = AF_UNIX;
+   cb_sockaddr = (struct sockaddr*)addr;
+   cb_sockaddr_len = SUN_LEN(addr);
+   return 0;
+}
+
+int km_init_started_callback(char* cb)
+{
+   char* cb_copy = strdup(cb);
+   char* savptr = NULL;
+   char* delim = ":";
+
+   char* cb_proto = strtok_r(cb_copy, delim, &savptr);
+   if (cb_proto == NULL) {
+      km_warnx("strtok_r returned NULL. cb_copy=%s", cb_copy);
+      free(cb_copy);
+      return -1;
+   }
+   if (strcmp(cb_proto, "udp") == 0) {
+      char* cb_addr = strtok_r(NULL, delim, &savptr);
+      char* cb_port = strtok_r(NULL, delim, &savptr);
+      km_warnx("callback addr: %s  port:%s", cb_addr, cb_port);
+      if (km_build_ipaddr(cb_addr, cb_port) < 0) {
+         free(cb_copy);
+         return -1;
+      }
+      free(cb_copy);
+      cb_domain = AF_INET;
+      cb_type = SOCK_DGRAM;
+      return 0;
+   } else if (strcmp(cb_proto, "tcp") == 0) {
+      km_warnx("callbacks on tcp sockets - not yet");
+      free(cb_copy);
+      return -1;
+   } else if (strcmp(cb_proto, "unix") == 0) {
+      char* cb_path = strtok_r(NULL, "", &savptr);
+      if (km_build_unixaddr(cb_path) < 0) {
+         free(cb_copy);
+         return -1;
+      }
+      free(cb_copy);
+      cb_domain = AF_UNIX;
+      cb_type = SOCK_SEQPACKET;
+
+      if ((cb_socket = km_internal_socket(cb_domain, cb_type, 0)) < 0) {
+         km_warn("cannnot open callback socket");
+         return -1;
+      }
+      if (connect(cb_socket, cb_sockaddr, cb_sockaddr_len) < 0) {
+         km_warn("cannnot open connect callback socket");
+         return -1;
+      }
+      return 0;
+   }
+   km_warnx("unrecognized callback type '%s' - only udp, tcp, unix", cb_proto);
+   free(cb_copy);
+   return -1;
+}
+
+static void km_fire_callback(int msg)
+{
+   if (cb_sockaddr == NULL) {
+      return;
+   }
+
+   if (cb_domain == AF_UNIX) {
+      if (send(cb_socket, &msg, sizeof(msg), 0) < 0) {
+         km_warn("km_fire_callback: send error");
+      }
+      return;
+   }
+
+   // cb_domain must == AF_INET
+   int sock = km_internal_socket(cb_domain, cb_type, 0);
+   if (sock < 0) {
+      km_warn("callback socket create");
+      return;
+   }
+
+   if (sendto(sock, &msg, sizeof(msg), 0, cb_sockaddr, cb_sockaddr_len) < 0) {
+      km_warn("callback udp sendto failed");
+   }
+   close(sock);
+}
+
+void km_fire_km_started_callback()
+{
+   km_fire_callback(0x304d4b);
```

Comment:
Where do these hex values come from?

0x304d4b
0x314d4b
0x324d4b

Reply:
"KM0". "KM1". "KM2" in ascii...